### PR TITLE
Latency Improvements and Memory Updates

### PR DESCRIPTION
**Pull request by marilynzhang**
Latency Improvements
- Parallelize shape context descriptor distance calculation step by using multiprocessing. This reduces our latency by about 0.5s. (Current latency with algorithm running HDBSCAN is now around 2.2s and with DBSCAN is around 1.5s, down from 3s/1.8s on average respectively.)
- Calculate median latency alongside mean

Memory Updates
- Instead of calculating the total memory used by the process during the icon finder function's execution (which will yield large memory usage -- 4000 MiBs -- for our large dataset and smaller memory usage -- 650 MiBs -- for our small dataset), calculate the *additional auxiliary memory* required by the function during execution. In particular, this will not include the memory needed to store the image. The auxiliary memory used by the algorithm is < 20 MiBs on average/median.
- Calculate median memory usage alongside mean

### CODE REVIEW

File: modules/clustering_algorithms.py [hunk 1]
```diff
@@ -13,7 +13,7 @@ class SklearnClusterer(abc.ABC):
     clusterer: object of type sklearn.cluster. All sklearn.cluster objects
      follow the same API (ie, you can call .fit(X) on them).
   """
-  clusterer: sklearn.cluster
+  clusterer = None
```
Comment (ewadkins) [inline modules/clustering_algorithms.py:16]: What's the reasoning behind removing the type? If this is necessary, can you also then update the attributes pydoc explaining the type better.

File: modules/benchmark_pipeline.py [hunk 2]
```diff
@@ -267,8 +278,9 @@ def evaluate(
       self.image_list, self.gold_boxes = analysis_util.scale_images_and_bboxes(
           self.image_list, self.gold_boxes, 5, 5)
 
-    avg_runtime_secs, avg_memory_mbs = self.find_icons(icon_finder_object,
-                                                       output_path, True, True)
+    avg_runtime_secs, avg_memory_mbs = self.find_icons(
+        icon_finder_object, output_path, True,
+        True)
```
Comment (ewadkins) [inline modules/benchmark_pipeline.py:283]: I know this is existing code, but can you add the arg names (calc_latency and calc_memory) to the two booleans here? Easier to see what the values do, and also is less error prone if the ordering of the args is ever switched for some reason.

File: modules/icon_finder_shape_context.py [hunk 3]
```diff
@@ -99,16 +122,25 @@ def _get_similar_contours(
...
+    nearby_contours = [
+        contour_and_distance[0]
+        for contour_and_distance in contours_and_distances
+        if contour_and_distance is not None
+    ]
+    nearby_distances = [
+        contour_and_distance[1]
+        for contour_and_distance in contours_and_distances
+        if contour_and_distance is not None
+    ]
```
Comment (ewadkins) [inline modules/icon_finder_shape_context.py:143]: Could do something like:
```
nearby_contours, nearby_distances = zip(*[x for x in contours_and_distances if x is not None])
```

File: modules/icon_finder_shape_context.py [hunk 4]
```diff
@@ -78,17 +103,15 @@ def _get_similar_contours(
...
-
     icon_pointset = algorithms.resize_pointset(icon_contour_keypoints,
                                                self.sc_min_num_points,
                                                self.sc_max_num_points,
                                                icon_contour_nonkeypoints)
     # expand the 1st dimension so that the shape is (n, 1, 2),
     # which is what shape context algorithm wants
     icon_contour_3d = np.expand_dims(icon_pointset, axis=1)
-
+    pool = multiprocessing.Pool(None)
```
Comment (ewadkins) [inline modules/icon_finder_shape_context.py:113]: (nit) Add a comment noting that this uses as many processes as available CPUs?

File: modules/icon_finder_shape_context.py [hunk 5]
```diff
@@ -52,6 +54,29 @@ def __init__(self,
...
+    """Calculate distance between icon and image contour.
+
+    Arguments:
+        icon_contour_3d: icon contour in shape context's format
+        image_contour_3d: image contour in shape context's format
+        (n, 1, 2)
+
+    Returns:
+        (distance, image_contour_3d) if there weren't any exceptions;
+        otherwise, does not return anything
```
Comment (ewadkins) [inline modules/icon_finder_shape_context.py:68]: Something like `(distance, image_contour_3d), or None if there was an exception` might be better.

**Review by ewadkins — CHANGES_REQUESTED** [review]

File: modules/clustering_algorithms.py (hunk 1 above)
Comment (marilynzhang) [reply modules/clustering_algorithms.py:16]: Updated with the correct base type now! (sklearn.cluster is actually not a type, just a module)

File: modules/icon_finder_shape_context.py [hunk 6]
```diff
@@ -99,16 +123,17 @@ def _get_similar_contours(
...
+                       args=(
+                           icon_contour_3d,
+                           image_contour_3d,
+                       ),
+                       callback=contours_and_distances.append)
+
+    pool.close()
+    pool.join()
+    nearby_contours, nearby_distances = zip(
+        *list(filter(None, contours_and_distances)))
```
Comment (ewadkins) [inline modules/icon_finder_shape_context.py:136]: Nice, I didn't even realize None could be passed to `filter` to filter out anything that evaluates to false.

File: modules/benchmark_pipeline.py [hunk 7]
```diff
@@ -31,7 +31,8 @@ def __init__(self, tfrecord_path: str = defaults.TFRECORD_PATH):
 
     # ----------------------the below are set by algorithm --------------------
     self.proposed_boxes = []  # proposed lists of bounding boxes for each image
-    self.image_clusters = []  # list of each image's contour clusters (analysis)
+    self.image_clusters = [
+    ]  # list of each image's contour clusters (analysis)
```
Comment (ewadkins) [inline modules/benchmark_pipeline.py:35]: (super nit) Can you add the comments on the line before, so you don't get any weird breaks like this.

File: modules/types.py [hunk 8]
```diff
@@ -1,6 +1,7 @@
 """This modules contains custom pytype types for other modules.
 """
 
-from typing import Union
+from typing import Tuple, Union
 
 OptionalFloat = Union[None, float]
+OptionalTuple = Union[None, Tuple]
```
Comment (ewadkins) [inline modules/types.py:7]: Hmm, can we use the actual `Optional` type here? It accomplishes the same thing. Also, if it's not too much of a hassle, you could change the other OptionalXXX[...] types to Optional[XXX[...]] pretty much with a find+replace.

Comment (ewadkins) [reply modules/types.py:7]: I should have asked for this change in previous PRs :)

**Review by ewadkins — APPROVED** [review]
Pls address the comments before submitting, thanks!

File: modules/types.py (hunk 8 above)
Comment (swongly) [reply modules/types.py:7]: Nice.  I probably suggested this.  Lol.  https://docs.python.org/3/library/typing.html#typing.Optional

File: modules/icon_finder_shape_context.py [hunk 9]
```diff
@@ -52,6 +54,28 @@ def __init__(self,
...
+        image_contour_3d: image contour in shape context's format
+        (n, 1, 2)
+
+    Returns:
+        (distance, image_contour_3d), or None if there was an exception
+    """
+    try:
+      distance = algorithms.shape_context_distance(icon_contour_3d,
+                                                   image_contour_3d)
+      if distance < self.sc_distance_threshold:
```
Comment (swongly) [inline modules/icon_finder_shape_context.py:72]: Do we not care if distance >= self.sc_distance_threshold?

Comment (marilynzhang) [reply modules/icon_finder_shape_context.py:72]: That's correct - we only consider clusters whose distances is below a lenient threshold

File: modules/types.py (hunk 8 above)
Comment (marilynzhang) [reply modules/types.py]: Woah nice, didn't know an actual ```Optional``` existed! And good memory Stan :) - https://github.com/googleinterns/acuiti/pull/24#discussion_r472369904

**Pull request merged**